File: java/src/main/java/com/showmeyourcode/playground/java/code/exercise/recursion/FibonacciExercise.java
package com.showmeyourcode.playground.java.code.exercise.recursion;

/**
 * Calculate n Fibonacci sequence elements using recursion and tail recursion.
 */
public class FibonacciExercise {

    private FibonacciExercise() {
    }

    public static int fibonacci(int n) {
        if (n == 0) {
            return 0; // Base case: fibonacci(0) = 0
        } else if (n == 1) {
            return 1; // Base case: fibonacci(1) = 1
        } else {
            return fibonacci(n - 1) + fibonacci(n - 2); // Recursive case
        }
    }

    public static int fibonacciTailRecursion(int n) {
        return fibonacciHelper(n, 0, 1);
    }

    private static int fibonacciHelper(int n, int number1, int number2) {
        if (n == 0) {
            return number1; // Base case: return the nth Fibonacci number
        } else {
            return fibonacciHelper(n - 1, number2, number1 + number2); // Tail recursive call
        }
    }
}

File: kotlin/src/main/kotlin/com/showmeyourcode/playground/kotlin/code/exercise/recursion/FibonacciExercise.kt
package com.showmeyourcode.playground.kotlin.code.exercise.recursion

fun fibonacci(n: Int): Int {
    return when (n) {
        0 -> 0
        1 -> 1
        else -> fibonacci(n - 2) + fibonacci(n - 1)
    }
}

fun fibonacciTailRecursion(
    n: Int,
    a: Int = 0,
    b: Int = 1
): Int {
    return when (n) {
        0 -> a
        1 -> b
        else -> fibonacciTailRecursion(n - 1, b, a + b)
    }
}
